react/components/Hydration/PreventHydration.tsx

```diff
@@ -152,6 +106,49 @@ const useAssetLoading = (extensionPointId?: string, shouldLoad?: boolean) => {
   return { isLoaded }
 }
 
+/** Prevents images from turning lazy again after re-rendering the component,
+ * to prevent them from flickering. */
+const useEagerImages = (treePath: string, shouldMakeImagesEager?: boolean) => {
+  const alreadyLoadedImages = useRef([] as HTMLImageElement[])
+  if (shouldMakeImagesEager) {
+    /**  Stores which descendant images were already lazyloaded.
+     * Those images will be set to load eagerly down below on the useLayoutEffect hook.
+     * This prevents them from blinking. */
+    const images = safeQuerySelectorAll<HTMLImageElement>(
+      `[data-hydration-id="${treePath}"] img.lazyloaded`
+    )
+
+    alreadyLoadedImages.current = Array.from(images)
+  }
+
+  useLayoutEffect(() => {
+    if (!shouldMakeImagesEager) {
+      return
+    }
+
+    const images = alreadyLoadedImages.current.map((image) =>
+      safeQuerySelector<HTMLImageElement>(`img[data-src="${image.src}"]`)
+    )
+
+    // Just a sanity check
+    if (!images) {
+      return
+    }
+
+    for (const image of images) {
+      if (!image) {
+        return
```

Comment:
```suggestion
        continue
```

Writing this down here before I forget